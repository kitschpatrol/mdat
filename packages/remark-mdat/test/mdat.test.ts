import remarkMdat, { type MdatCleanOptions, type Options, mdatClean, mdatSplit } from '../src'
import testRules from './assets/test-rules'
// @ts-expect-error - Intentionally invalid for testing purposes
import testRulesInvalid from './assets/test-rules-invalid'
import { type Root } from 'mdast'
import fs from 'node:fs/promises'
import { remark } from 'remark'
import remarkGfm from 'remark-gfm'
import { type VFile } from 'vfile'
import { describe, expect, it } from 'vitest'

async function expandStringToString(markdown: string, options: Options): Promise<string> {
	const result = await remark().use(remarkGfm).use(remarkMdat, options).process(markdown)
	return result.toString()
}

// Export for linter
export async function cleanString(markdown: string, options: MdatCleanOptions): Promise<string> {
	const result = await remark()
		.use(remarkGfm)
		.use(
			() =>
				// eslint-disable-next-line unicorn/consistent-function-scoping
				function (tree: Root, file: VFile) {
					mdatSplit(tree, file)
					mdatClean(tree, file, options)
				},
		)
		.process(markdown)
	return result.toString()
}

async function expandFileToString(file: string, options: Options): Promise<string> {
	const buffer = await fs.readFile(file)
	const result = await remark().use(remarkGfm).use(remarkMdat, options).process(buffer)
	return result.toString()
}

describe('comment expansion', () => {
	it('should expand comments', async () => {
		const expandedString = await expandFileToString('./test/assets/test-document.md', {
			rules: testRules,
		})
		expect(expandedString.toString()).toMatchSnapshot()
	})

	it('should be idempotent', async () => {
		const firstPass = await expandFileToString('./test/assets/test-document.md', {
			rules: testRules,
		})

		const secondPass = await expandStringToString(firstPass, {
			rules: testRules,
		})

		expect(firstPass.toString()).toEqual(secondPass.toString())
	})

	it('should expand prefixed comments only', async () => {
		const expandedString = await expandFileToString('./test/assets/test-document.md', {
			keywordPrefix: 'mm-',
			rules: testRules,
		})

		expect(expandedString.toString()).toMatchSnapshot()
	})

	it('should include the meta tag if asked', async () => {
		const expandedString = await expandFileToString('./test/assets/test-document.md', {
			addMetaComment: true,
			rules: testRules,
		})

		expect(expandedString.toString()).toMatchSnapshot()
	})

	it('should throw an error if rule set is invalid', async () => {
		await expect(
			expandFileToString('./test/assets/test-document.md', {
				rules: testRulesInvalid,
			}),
		).rejects.toThrow()
	})
})

describe('keyword case sensitivity', () => {
	it('should treat comment expansion keywords as case sensitive', async () => {
		const markdown = `<!-- KEYWORD -->\n<!-- kEyWoRd -->\n<!-- keyword -->\n`
		const options: Options = {
			rules: {
				// eslint-disable-next-line @typescript-eslint/naming-convention
				KEYWORD: "I'm yelling",
				kEyWoRd: "I'm emotional",
				keyword: "I'm basic",
			},
		}
		const expandedString = await expandStringToString(markdown, options)
		expect(expandedString.toString()).toMatchInlineSnapshot(`
			"<!-- KEYWORD -->

			I'm yelling

			<!-- /KEYWORD -->

			<!-- kEyWoRd -->

			I'm emotional

			<!-- /kEyWoRd -->

			<!-- keyword -->

			I'm basic

			<!-- /keyword -->
			"
		`)
	})
})

describe('compound rule handling', () => {
	it('should expand compound rules', async () => {
		const markdown = `<!-- compoundKeyword -->\n`
		const options: Options = {
			rules: {
				compoundKeyword: ['one', 'two', 'three'],
			},
		}
		const expandedString = await expandStringToString(markdown, options)
		expect(expandedString.toString()).toMatchInlineSnapshot(`
			"<!-- compoundKeyword -->

			one

			two

			three

			<!-- /compoundKeyword -->
			"
		`)
	})

	it('should pass option arrays to compound rules', async () => {
		const markdown = `<!-- compound [{option: 'yes'}, {option: 'it'}, {option: 'can'}] -->\n`
		const options: Options = {
			rules: {
				compound: [
					(options) => `My option is: ${(options as { option: string })?.option}`,
					(options) => `My option is: ${(options as { option: string })?.option}`,
					(options) => `My option is: ${(options as { option: string })?.option}`,
				],
			},
		}
		const expandedString = await expandStringToString(markdown, options)
		expect(expandedString.toString()).toMatchInlineSnapshot(`
			"<!-- compound [{option: 'yes'}, {option: 'it'}, {option: 'can'}] -->

			My option is: yes

			My option is: it

			My option is: can

			<!-- /compound -->
			"
		`)
	})
})
