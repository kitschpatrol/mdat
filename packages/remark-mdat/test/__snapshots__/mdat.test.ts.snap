// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment expansion > should expand comments 1`] = `
"# The MDAT sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

## Basic comment clean up

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm-\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.ts

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

ðŸŒ³ I am between two ferns ðŸŒ³

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, MDAT will report the error.

<!-- basic-empty -->

<!-- /basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and MDAT will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered-2 -->

I had to be last

<!-- /basic-ordered-2 -->

<!-- basic-ordered-1 -->

I had to be first

<!-- /basic-ordered-1 -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->test-rules.ts<!-- /basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

I am a paragraph 1

I am a paragraph 2

I am a paragraph 3

<!-- /basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->
>
> I am a paragraph 1
>
> I am a paragraph 2
>
> I am a paragraph 3
>
> <!-- /basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><b>Absolutely</b><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options -->

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->

## Adjacent top-level expansions with leading space

&#x20;<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->

## Expansion with leading space

&#x20;<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->
"
`;

exports[`comment expansion > should expand prefixed comments only 1`] = `
"# The MDAT sample document

## Basic comment expansion

<!-- basic -->

## Basic comment clean up

<!-- basic -->

Stale content that will be replaced

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm-\` option is set:

<!-- mm-basic -->

**A bold statement from test-rules.ts**

<!-- /mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

## Expansions with arguments, syntax forgiveness

<!--- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, MDAT will report the error.

<!-- basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and MDAT will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered-2 -->

<!-- basic-ordered-1 -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

<!-- basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic --><!-- /basic --><!-- basic --><!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic --><!-- basic -->Z

## Adjacent top-level expansions

<!-- basic --><b>Absolutely</b><!-- basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Adjacent top-level expansions

<!-- basic --><!-- basic -->

## Adjacent top-level expansions with leading space

&#x20;<!-- basic --><!-- basic -->

## Expansion with leading space

&#x20;<!-- basic -->
"
`;

exports[`comment expansion > should include the meta tag if asked 1`] = `
"<!--+ Warning: Content inside HTML comment blocks was generated by mdat and may be overwritten. +-->

# The MDAT sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

## Basic comment clean up

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm-\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.ts

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

ðŸŒ³ I am between two ferns ðŸŒ³

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, MDAT will report the error.

<!-- basic-empty -->

<!-- /basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and MDAT will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered-2 -->

I had to be last

<!-- /basic-ordered-2 -->

<!-- basic-ordered-1 -->

I had to be first

<!-- /basic-ordered-1 -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->test-rules.ts<!-- /basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

I am a paragraph 1

I am a paragraph 2

I am a paragraph 3

<!-- /basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->
>
> I am a paragraph 1
>
> I am a paragraph 2
>
> I am a paragraph 3
>
> <!-- /basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

<!-- basic -->

**A bold statement from test-rules.ts**

<!-- /basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><b>Absolutely</b><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options -->

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->

## Adjacent top-level expansions with leading space

&#x20;<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic --><!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->

## Expansion with leading space

&#x20;<!-- basic -->**A bold statement from test-rules.ts**<!-- /basic -->
"
`;
